wicketd/src/update_tracker.rs

```diff
@@ -251,31 +250,29 @@ impl UpdateTracker {
         UploadTrampolinePhase2ToMgs { status: status_rx, task }
     }
 
-    pub(crate) async fn update_log(
-        &self,
-        sp: SpIdentifier,
-    ) -> crate::update_events::UpdateLog {
+    pub(crate) async fn event_report(&self, sp: SpIdentifier) -> EventReport {
         let mut sp_update_data = self.sp_update_data.lock().await;
         match sp_update_data.entry(sp) {
-            Entry::Vacant(_) => crate::update_events::UpdateLog::default(),
+            Entry::Vacant(_) => EventReport::default(),
             Entry::Occupied(slot) => {
-                slot.get().update_log.lock().unwrap().clone()
+                slot.get().event_buffer.lock().unwrap().generate_report()
             }
         }
     }
 
     /// Clone the current state of the update log for every SP, returning a map
     /// suitable for conversion to JSON.
-    pub(crate) async fn update_log_all(
+    pub(crate) async fn event_report_all(
         &self,
-    ) -> BTreeMap<SpType, BTreeMap<u32, crate::update_events::UpdateLog>> {
+    ) -> BTreeMap<SpType, BTreeMap<u32, EventReport>> {
         let sp_update_data = self.sp_update_data.lock().await;
         let mut converted_logs = BTreeMap::new();
         for (sp, update_data) in &*sp_update_data {
-            let update_log = update_data.update_log.lock().unwrap().clone();
+            let event_report =
+                update_data.event_buffer.lock().unwrap().generate_report();
             let inner: &mut BTreeMap<_, _> =
                 converted_logs.entry(sp.type_).or_default();
-            inner.insert(sp.slot, update_log);
+            inner.insert(sp.slot, event_report);
         }
         converted_logs
```

Comment:
A thing these methods do not do is reset the event report once it's delivered. It looks like the old update log implementation didn't do that either. We either need to find a way to run some code after a successful request, or provide a way for wicket to tell wicketd/the event buffer which events it has already seen, and an event buffer to diff against that.

Reply:
That was intentional: we don't assume a single `wicket` invocation will be alive the entire time an update is running, so if a new `wicket` starts it could always fetch the full update log at any point. I think with your option 2 that would still work: a `wicket` that has seen some state before could request just new events, and a new `wicket` could request the full report.

I _think_ this is an optimization we could delay for now, though, right? The old update log had a pretty small max size - the `EventReport` is bigger but maybe not too big?

Reply:
Yeah, this is not a relevant optimization for now. However, for other reasons I implemented option 2 so we can plug this in at any time.